shell/browser/electron_browser_handler_impl.cc

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2019 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/electron_browser_handler_impl.h"
+
+#include <utility>
+
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/render_process_host.h"
+#include "mojo/public/cpp/bindings/self_owned_receiver.h"
+
+namespace electron {
+ElectronBrowserHandlerImpl::ElectronBrowserHandlerImpl(
+    content::RenderFrameHost* frame_host)
+    : render_process_id_(frame_host->GetProcess()->GetID()),
+      render_frame_id_(frame_host->GetRoutingID()),
+      weak_factory_(this) {
+  content::WebContents* web_contents =
+      content::WebContents::FromRenderFrameHost(frame_host);
+  content::WebContentsObserver::Observe(web_contents);
+}
+
+ElectronBrowserHandlerImpl::~ElectronBrowserHandlerImpl() = default;
+
+void ElectronBrowserHandlerImpl::BindTo(
+    mojo::PendingReceiver<mojom::ElectronBrowser> receiver) {
+  // Note: BindTo might be called for multiple times.
+  if (receiver_.is_bound())
+    receiver_.reset();
+
+  receiver_.Bind(std::move(receiver));
+  receiver_.set_disconnect_handler(base::BindOnce(
+      &ElectronBrowserHandlerImpl::OnConnectionError, GetWeakPtr()));
+}
+
+void ElectronBrowserHandlerImpl::WebContentsDestroyed() {
+  delete this;
+}
+
+void ElectronBrowserHandlerImpl::OnConnectionError() {
+  if (receiver_.is_bound())
+    receiver_.reset();
```

Comment:
Given we never expect to receive another message in the future after this method is called, does it make sense to `delete this` here instead?

Reply:
@nornagon don't we need to reset the receiver to avoid errors like:
```
Check failed: !connected. ElectronBrowser::MessageSyncCallback was destroyed without first either being run or its corresponding binding being closed. It is an error to drop response callbacks which still correspond to an open interface pipe.
```
?

Reply:
The receiver will automatically be reset when it's deleted.